core/trino-main/src/main/java/io/trino/node/AnnounceNodeAnnouncer.java

```diff
@@ -0,0 +1,182 @@
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.errorprone.annotations.FormatMethod;
+import com.google.inject.Inject;
+import io.airlift.http.client.HttpClient;
+import io.airlift.http.client.Request;
+import io.airlift.http.client.StaticBodyGenerator;
+import io.airlift.http.client.StatusResponseHandler.StatusResponse;
+import io.airlift.log.Logger;
+
+import java.net.URI;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static io.airlift.http.client.HttpUriBuilder.uriBuilderFrom;
+import static io.airlift.http.client.Request.Builder.preparePost;
+import static io.airlift.http.client.StaticBodyGenerator.createStaticBodyGenerator;
+import static io.airlift.http.client.StatusResponseHandler.createStatusResponseHandler;
+import static jakarta.ws.rs.core.MediaType.TEXT_PLAIN;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+public class AnnounceNodeAnnouncer
+        implements Announcer
+{
+    private static final Logger log = Logger.get(AnnounceNodeAnnouncer.class);
+
+    private final HttpClient httpClient;
+    private final List<URI> announceUris;
+    private final ScheduledExecutorService announceExecutor;
+    private final StaticBodyGenerator currentHostAnnouncement;
+    private final AtomicLong lastWarningLogged = new AtomicLong(0);
+
+    private final AtomicReference<State> state = new AtomicReference<>(State.NOT_STARTED);
+
+    private enum State
+    {
+        NOT_STARTED,
+        STARTED,
+        STOPPED,
+    }
+
+    @Inject
+    public AnnounceNodeAnnouncer(InternalNode currentNode, AnnounceConfig config, @ForAnnouncer HttpClient httpClient)
+    {
+        this(currentNode.getInternalUri(), config.getCoordinatorUris(), httpClient);
+    }
+
+    @VisibleForTesting
+    public AnnounceNodeAnnouncer(URI internalUri, Collection<URI> coordinatorUris, HttpClient httpClient)
+    {
+        URI currentUri = internalUri;
+        this.announceUris = coordinatorUris.stream()
+                .filter(uri -> !currentUri.equals(uri))
+                .map(uri -> uriBuilderFrom(uri).appendPath("/v1/announce").build())
+                .distinct()
+                .toList();
+        this.httpClient = requireNonNull(httpClient, "httpClient is null");
+        this.announceExecutor = newSingleThreadScheduledExecutor(daemonThreadsNamed("node-state-poller-%s"));
```

Comment:
No need for `-%s` with `newSingleThreadScheduledExecutor` since there is only one thread